nexus/codec.go

```diff
@@ -0,0 +1,204 @@
+			}
+			return nil, err
+		}
+		return p, nil
+	}
+	return nil, errCodecIncompatible
+}
+
+func (c codecChain) Decode(s *Stream, v any) error {
+	lenc := len(c)
+	for i := range c {
+		l := c[lenc-i-1]
+		if err := l.Decode(s, v); err != nil {
+			if errors.Is(err, errCodecIncompatible) {
+				continue
+			}
+			return err
+		}
+		return nil
+	}
+	return errCodecIncompatible
+}
+
+var _ Codec = codecChain{}
+
+type jsonCodec struct{}
+
+func (jsonCodec) Decode(s *Stream, v any) error {
+	if !isMediaTypeJSON(s.Header["content-type"]) {
+		return errCodecIncompatible
+	}
+	body, err := io.ReadAll(s.Reader)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(body, &v)
+}
+
+func (jsonCodec) Encode(v any) (*Stream, error) {
+	body, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+	return &Stream{
+		Header: map[string]string{
+			"content-type":   "application/json",
+			"content-length": fmt.Sprintf("%d", len(body)),
+		},
+		Reader: io.NopCloser(bytes.NewReader(body)),
+	}, nil
+}
+
+var _ Codec = jsonCodec{}
+
+type nilCodec struct{}
+
+func (nilCodec) Decode(s *Stream, v any) error {
+	if s.Header["content-length"] != "0" {
+		return errCodecIncompatible
+	}
+	rv := reflect.ValueOf(v)
+	if rv.Kind() != reflect.Pointer {
+		return fmt.Errorf("could not decode non pointer value: %v", v)
+	}
+	re := rv.Elem()
+	if !re.CanSet() {
+		return fmt.Errorf("non settable type: %v", v)
+	}
+	// Set the zero value for the given type.
+	re.Set(reflect.Zero(re.Type()))
+
+	return nil
+}
+
+func (nilCodec) Encode(v any) (*Stream, error) {
+	if v != nil {
+		rv := reflect.ValueOf(v)
+		if !(rv.Kind() == reflect.Ptr && rv.IsNil()) {
+			return nil, errCodecIncompatible
+		}
+	}
+	return &Stream{
+		Header: map[string]string{"content-length": "0"},
+		Reader: nil,
+	}, nil
+}
+
+var _ Codec = nilCodec{}
+
+type byteSliceCodec struct{}
+
+func (byteSliceCodec) Decode(s *Stream, v any) error {
+	if !isMediaTypeOctetStream(s.Header["content-type"]) {
+		return errCodecIncompatible
+	}
+	if bPtr, ok := v.(*[]byte); ok {
+		if bPtr == nil {
+			return fmt.Errorf("cannot decode into nil pointer: %v", v)
+		}
+		b, err := io.ReadAll(s.Reader)
+		if err != nil {
+			return err
+		}
+		*bPtr = b
+		return nil
+	}
+	// v is *any
+	rv := reflect.ValueOf(v).Elem()
+	if rv.Kind() == reflect.Interface {
+		b, err := io.ReadAll(s.Reader)
+		if err != nil {
+			return err
+		}
+		rv.Set(reflect.ValueOf(b))
+		return nil
+	}
+	return fmt.Errorf("unsupported value type for content: %v", v)
+}
+
+func (byteSliceCodec) Encode(v any) (*Stream, error) {
+	if b, ok := v.([]byte); ok {
+		return &Stream{
+			Header: map[string]string{
+				"content-type":   "application/octet-stream",
+				"content-length": fmt.Sprintf("%d", len(b)),
+			},
+			Reader: io.NopCloser(bytes.NewReader(b)),
+		}, nil
+	}
+	return nil, errCodecIncompatible
+}
+
+var _ Codec = byteSliceCodec{}
+
+type compositeCodec struct {
+	codecChain
+}
+
+var defaultCodec Codec = compositeCodec{
```

Comment:
```suggestion
var DefaultCodec Codec = compositeCodec{
```

Should expose this IMO

Reply:
Let's leave this unexposed for now until it's deemed necessary.